# Задание 8.11.2022
---

In [7]:
import pandas as pd
import numpy as np

Загрузка данных из orders.csv и customers.csv

In [8]:
orders = pd.read_csv('orders.csv', sep=',', index_col='id')
customers = pd.read_csv('customers.csv', sep=',', index_col='id')

## 1. Сколько заказов, отправлено первым классом за последние 3 лет?

In [19]:
orders['order_date'] = pd.to_datetime(orders['order_date'])
max_date = orders.query('ship_mode == "First"')['order_date'].max()
from_date = pd.Timestamp(year=max_date.year-3, month=max_date.month, day=max_date.day)
orders.query('ship_mode == "First" & order_date >= @from_date')['sales'].count()

645

**Ответ**: 645

## 2. Сколько в базе клиентов из Калифорнии?

In [20]:
customers.query('state == "California"')['name'].count()

161

**Ответ**: 161

## 3. Сколько заказов они сделали?

In [21]:
orders_with_customers = orders.merge(customers, how='inner', left_on='customer_id', right_index=True)
orders_with_customers.query('state == "California"')['sales'].count()

1006

**Ответ**: 1006

## 4. Постройте сводную таблицу средних чеков по всем штатам за каждый год.

In [23]:
orders_with_customers['order_date_year'] = orders_with_customers['order_date'].apply(lambda x: x.year)
orders_with_customers.pivot_table(values='sales', index='state', columns='order_date_year', aggfunc=np.mean)

order_date_year              2014         2015        2016        2017
state                                                                 
Alabama                425.521556   501.367650  854.347188  417.962833
Arizona                475.161875   703.277355  699.062970  535.657542
Arkansas              1098.006000    55.830000  103.054000  214.764800
California             563.671775   404.165555  426.373201  422.228417
Colorado               351.093136   433.258312  540.723206  365.629421
Connecticut            954.951600   183.639600  392.534000  479.202176
Delaware               341.779273   401.070750  304.878545  894.523583
District of Columbia   267.171333          NaN  465.645333         NaN
Florida                445.699714   333.154130  318.362135  318.330974
Georgia                462.506500   559.751000  521.909906  390.538583
Illinois               466.230380   352.518270  407.708433  315.350126
Indiana                351.278125   198.633400  258.387833  277.871879
Iowa                   504.827000  1331.897667  529.363750  544.059000
Kansas                        NaN   237.055000   91.305333  979.626000
Kentucky               438.823308   477.137333  264.063059  370.057071
Louisiana              478.414200   364.051857  252.510571  414.785500
Maryland                11.703000   342.305000   63.653000   78.272750
Massachusetts          632.285182   285.898875  899.047273  317.832091
Michigan               444.491700   827.384654  521.852258  383.253493
Minnesota              521.840974   585.269167  552.985417  534.553833
Mississippi            673.085200  2171.503000  435.542556  482.433125
Missouri               276.453000   491.229000  469.916000  452.615111
Nebraska               465.432667   261.397200  120.009333  857.274500
Nevada                  22.605333    56.844000   58.580000  487.451000
New Hampshire          169.915600  1420.661500  898.574000  246.100000
New Jersey             239.149800   268.907000  276.567591  453.319333
New Mexico             341.086400   181.444667  371.928800  276.682571
New York               436.122764   442.318855  508.637979  583.497825
North Carolina        1099.830273   279.665798  769.887707  578.589109
Ohio                   298.662170   318.697911  313.263246  249.743436
Oklahoma                62.808000   538.920000  361.824000  689.723600
Oregon                 774.243750   977.450667  849.970500  290.465429
Pennsylvania           293.171032   556.985115  605.352251  391.808053
Rhode Island            87.067000   273.439333  659.479667  221.556154
South Dakota            21.190000   263.990000   15.210000  586.669333
Tennessee              526.849467   637.160000  200.831842  323.761131
Texas                  584.219643   343.701678  342.953301  420.732574
Utah                   293.355375   408.489333  626.938667  467.010571
Virginia              1048.344882   491.514667  466.331524  268.270781
Washington             601.211070   491.759512  660.959755  576.431940
Wisconsin              318.099000   831.201000  444.452824  797.464704